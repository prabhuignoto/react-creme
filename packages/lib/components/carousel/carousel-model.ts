import React from 'react';

export type CarouselProps = {
  autoPlay?: number;
  border?: boolean;
  children: React.ReactNode | React.ReactNode[];
  direction?: 'horizontal' | 'vertical';
  enableSwipe?: boolean;
  focusable?: boolean;
  height?: number;
  size?: 'sm' | 'md' | 'lg';
  transition?: string;
};

export type CarouselItemProps = {
  height: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width: number;
};

export type CarouselItemsProps = Pick<
  CarouselProps,
  'children' | 'direction' | 'height' | 'size'
> & {
  activePage: number;
  carouselItems: CarouselItemProps[];
  totalItems: number;
  width: number;
};

export type CarouselTrackProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  activeIndex: number;
  focusable?: boolean;
  handleSelection: (index: number) => void;
  hideNext?: boolean;
  hidePrevious?: boolean;
  length: number;
  onNext: () => void;
  onPrevious: () => void;
};

export type CarouselButtonProps = Pick<
  CarouselProps,
  'direction' | 'focusable' | 'size'
> & {
  hide?: boolean;
  label: string;
  onClick: () => void;
  position: 'left' | 'right';
};
