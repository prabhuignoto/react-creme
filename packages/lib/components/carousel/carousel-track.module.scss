@use '../../design/core.scss';
@use '../../design/list.scss';
@use '../../design/theme.scss';

.track {
  &.track_horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.track_vertical {
    @extend %list;
    width: 3rem;
  }
}

.track_item {
  border: 1px solid theme.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.track_item:hover {
  cursor: pointer;
}

.track_item_selected {
  background: theme.$primary;
  border: 1px solid theme.$primary;
}

.icon {
  @extend %icon-md;
}

.track_wrapper_horizontal {
  @extend %center;
}

.track_wrapper_vertical {
  @extend %list;
}
