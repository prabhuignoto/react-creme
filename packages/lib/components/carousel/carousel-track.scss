@use '../../design/core.scss';
@use '../../design/list.scss';
@use '../../design/theme.scss';

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  border: 1px solid theme.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: theme.$primary;
  border: 1px solid theme.$primary;
}

.rc-carousel-icon {
  @extend %icon-md;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
