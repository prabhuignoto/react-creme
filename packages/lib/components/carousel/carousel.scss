@use '../../design/button.scss';
@use '../../design/icon.scss';
@use '../../design/theme.scss';
@use '../../design/list.scss';
@use '../../design/layout.scss';
@use '../../design/position.scss';

.rc-carousel-container {
  @extend %border;
  // @extend %shadow-inset;
  height: 100%;
  width: 100%;

  &.rc-carousel-container-horizontal {
    @extend %col;
  }

  &.rc-carousel-container-vertical {
    @extend %center;
  }
}

.rc-carousel-wrapper {
  overflow: hidden;
  position: relative;

  &.rc-carousel-wrapper-horizontal {
    height: calc(100% - 4rem);
    margin-top: 1rem;
    width: 95%;
  }

  &.rc-carousel-wrapper-vertical {
    height: calc(100% - 3rem);
    margin-left: 1rem;
    width: calc(100% - 3rem);
  }
}

.rc-carousel-track-wrapper {
  @extend %center;

  &.rc-carousel-track-wrapper-horizontal {
    height: 3rem;
    width: 100%;
  }

  &.rc-carousel-track-wrapper-horizontal {
    width: 3rem;
  }
}

.rc-carousel {
  height: 100%;
  left: 0;
  list-style: none;
  margin: 0;
  padding: 0;
  position: absolute;
  top: 0;
  transition: transform 0.5s var(--transition);
}

.rc-carousel-item {
  @extend %center;
  @extend %border-radius;
  height: 100%;
  position: absolute;
  top: 50%;

  img {
    @extend %border-radius;
    max-height: 100%;
    max-width: 100%;
  }
}

.rc-carousel-item-container {
  @extend %center;
  height: 95%;
  width: 95%;
}

.rc-carousel-btn {
  border-radius: 50%;

  &.rc-carousel-btn-horizontal {
    margin: 0 1rem;
  }

  &.rc-carousel-btn-vertical {
    margin: 1rem 0;
  }

  &.rc-carousel-btn-hide {
    visibility: hidden;
  }
}

.rc-carousel-btn-left {
  transform: rotate(-180deg);
}

.rc-carousel-btn-left:not(.rc-carousel-btn-horizontal) {
  transform: rotate(-90deg);
}

.rc-carousel-btn-right:not(.rc-carousel-btn-horizontal) {
  transform: rotate(90deg);
}
