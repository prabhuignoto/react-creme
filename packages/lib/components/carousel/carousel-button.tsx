import classNames from 'classnames';
import React from 'react';
import { ChevronRightIcon } from '../../icons';
import { Button } from '../button/button';
import { CarouselButtonProps } from './carousel-model';
import './carousel.scss';

const CarouselButton: React.FunctionComponent<CarouselButtonProps> = ({
  onClick,
  position,
  hide,
  direction,
  focusable,
  label,
}: CarouselButtonProps) => {
  return (
    <span
      className={classNames(
        [
          'rc-carousel-btn',
          `rc-carousel-btn-${position}`,
          `rc-carousel-btn-${direction}`,
        ],
        {
          'rc-carousel-btn-hide': hide,
        }
      )}
    >
      <Button type="icon" onClick={onClick} focusable={focusable} label={label}>
        <ChevronRightIcon />
      </Button>
    </span>
  );
};

CarouselButton.displayName = 'CarouselButton';

export { CarouselButton };
