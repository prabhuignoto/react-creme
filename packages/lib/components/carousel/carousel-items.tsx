import React, { CSSProperties, ReactNode, useMemo } from "react";
import { CarouselItemsModel } from "./carousel-model";
import "./carousel.scss";

const CarouselItems: React.FunctionComponent<CarouselItemsModel> = ({
  activePage = 0,
  carouselItems,
  children,
  direction,
  height,
  totalItems,
  width,
}: CarouselItemsModel) => {
  const carouselStyle = useMemo(() => {
    return {
      width: `${width * totalItems}px`,
      transform:
        direction === "horizontal"
          ? `translateX(-${width * activePage}px)`
          : `translateY(-${height * activePage}px)`,
    } as CSSProperties;
  }, [width, height, activePage]);

  return (
    <ul className={"rc-carousel"} style={carouselStyle} role="list">
      {carouselItems.map((item, index) => (
        <li
          key={item.id}
          className={"rc-carousel-item"}
          role="listitem"
          data-visible={activePage === index}
          aria-hidden={activePage !== index}
          style={{
            width: item.width,
            visibility: item.width === 0 ? "hidden" : "visible",
            top: direction === "horizontal" ? 0 : item.top,
            left: direction === "horizontal" ? item.left : 0,
          }}
        >
          <div className="rc-carousel-item-container">
            {(children as ReactNode[])[index]}
          </div>
        </li>
      ))}
    </ul>
  );
};

CarouselItems.displayName = "CarouselItems";

export { CarouselItems };
