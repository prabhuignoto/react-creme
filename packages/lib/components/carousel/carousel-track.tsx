import classNames from 'classnames';
import React, { useMemo } from 'react';
import { CarouselButton } from './carousel-button';
import { CarouselTrackProps } from './carousel-model';
import styles from './carousel-track.module.scss';
import rootStyles from './carousel.module.scss';

const CarouselTrack: React.FunctionComponent<CarouselTrackProps> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
  focusable,
}: CarouselTrackProps) => {
  const carouselTrackClass = useMemo(
    () =>
      classNames([
        styles.track,
        direction === 'horizontal'
          ? styles.track_horizontal
          : styles.track_vertical,
      ]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames([
        rootStyles.track_wrapper,
        direction === 'horizontal'
          ? styles.track_wrapper_horizontal
          : styles.track_wrapper_vertical,
        // [styles[`track-wrapper-${direction}`]]: true,
      ]),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
        focusable={focusable}
        label="Previous"
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              styles.track_item,
              index === activeIndex ? styles.track_item_selected : '',
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
        focusable={focusable}
        label="Next"
      />
    </div>
  );
};

export { CarouselTrack };
