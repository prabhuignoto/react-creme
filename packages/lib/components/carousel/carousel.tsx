import classNames from 'classnames';
import { nanoid } from 'nanoid';
import React from 'react';
import {
  CSSProperties,
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import ResizeObserver from 'resize-observer-polyfill';
import { useDebounce } from 'use-debounce';
import useSwipe from '../common/effects/useSwipe';
import { CarouselItems } from './carousel-items';
import { CarouselItemProps, CarouselProps } from './carousel-model';
import { CarouselTrack } from './carousel-track';
import styles from './carousel.module.scss';

// Carousel is a React Function Component for displaying a carousel of items.
const Carousel: React.FunctionComponent<CarouselProps> = ({
  autoPlay = 0,
  children,
  direction = 'horizontal',
  height = 400,
  transition = 'cubic-bezier(0.55, 0.08, 0.68, 0.53)',
  focusable = true,
  border = false,
  size = 'md',
}) => {
  // Carousel items state. Each item has an id, visibility status, and its dimensions.
  const [carouselItems, setCarouselItems] = useState<CarouselItemProps[]>(
    Array.isArray(children)
      ? children.map(() => ({
          height: 0,
          id: nanoid(),
          visible: false,
          width: 0,
        }))
      : []
  );

  // Ref to store the count of the carousel items.
  const trackCount = useRef(Array.isArray(children) ? children.length : 0);
  // Ref to store the autoPlay interval id.
  const autoPlayRef = useRef<number>();

  // State to store the index of the active (currently displayed) item.
  const [activePage, setActivePage] = useState(0);
  // State to store the dimensions of each slide.
  const [slideWidth, setSlideWidth] = useState(0);
  const [slideHeight, setSlideHeight] = useState(0);
  // State to store the auto-playing status of the carousel.
  const [isAutoPlaying, setAutoPlaying] = useState(!!autoPlay);

  // Debounced versions of slide dimensions for performance optimization.
  const [debouncedSlideWidth] = useDebounce(slideWidth, 100);
  const [debouncedSlideHeight] = useDebounce(slideHeight, 100);

  // Ref to store the instance of ResizeObserver for resizing the carousel items.
  const resizeObserver = useRef<ResizeObserver | null>(null);

  // Callback function to handle "Next" action in the carousel.
  const handleNext = useCallback(() => {
    activePage < trackCount.current - 1 && setActivePage(prev => prev + 1);
  }, [activePage]);

  // Callback function to handle "Previous" action in the carousel.
  const handlePrevious = useCallback(() => {
    setActivePage(prev => prev - 1);
  }, [activePage]);

  // Callback function to handle page activation in the carousel.
  const handleActivatePage = useCallback(
    (pageIndex: number) => setActivePage(pageIndex),
    []
  );

  // Callback function to handle the initial setup of the carousel.
  const onInitRef = useCallback((node: HTMLDivElement) => {
    if (node) {
      const { clientHeight, clientWidth } = node;
      setSlideWidth(clientWidth);
      setSlideHeight(clientHeight);

      resizeObserver.current = new ResizeObserver(() => {
        const { clientHeight, clientWidth } = node as HTMLElement;
        setSlideWidth(clientWidth);
        setSlideHeight(clientHeight);
      });

      resizeObserver.current.observe(node);
    }
  }, []);

  // Effect to update the carousel items whenever the dimensions change.
  useLayoutEffect(() => {
    if (slideWidth && slideHeight && Array.isArray(children)) {
      const prop = direction === 'horizontal' ? 'left' : 'top';
      setCarouselItems(
        children.map((_, index) => ({
          height: debouncedSlideHeight,
          id: nanoid(),
          [prop]: `${
            direction === 'horizontal'
              ? index * slideWidth
              : index * slideHeight
          }px`,
          visible: false,
          width: debouncedSlideWidth,
        }))
      );
    }
  }, [debouncedSlideWidth, debouncedSlideHeight]);

  // Memoized value to determine whether to hide the next button.
  const hideNextButton = useMemo(
    () => activePage === trackCount.current - 1,
    [activePage]
  );

  // Effect to handle auto-playing of the carousel.
  useLayoutEffect(() => {
    if (
      !autoPlayRef.current &&
      autoPlay &&
      activePage < carouselItems.length - 1
    ) {
      setAutoPlaying(true);
      autoPlayRef.current = window.setInterval(() => {
        setActivePage(page => {
          if (page === carouselItems.length - 1) {
            clearInterval(autoPlayRef.current);
            setAutoPlaying(false);
            return page;
          }

          return page + 1;
        });
      }, autoPlay);
    }
  }, [autoPlay, carouselItems.length]);

  // Memoized value to determine whether to hide the previous button.
  const hidePreviousButton = useMemo(() => activePage === 0, [activePage]);

  // Classnames for the carousel track and container.
  const carouselTrackClass = useMemo(
    () =>
      classNames([
        styles.track_wrapper,
        direction === 'horizontal' ? styles.track_wrapper_horizontal : '',
      ]),
    [isAutoPlaying]
  );
  const carouselContainerClass = useMemo(
    () =>
      classNames(
        [
          styles.container,
          direction === 'horizontal'
            ? styles.container_horizontal
            : styles.container_vertical,
        ],
        {
          'rc-carousel-border': border,
        }
      ),
    []
  );

  // Styles for the carousel wrapper.
  const carouselWrapperStyle = useMemo(
    () =>
      ({
        '--min-height': `${height}px`,
        '--transition': transition,
      }) as CSSProperties,
    []
  );

  // Classname for the carousel wrapper.
  const wrapperClass = useMemo(
    () =>
      classNames([
        styles.wrapper,
        direction === 'horizontal'
          ? styles.wrapper_horizontal
          : styles.wrapper_vertical,
      ]),
    []
  );

  // Swipe handler.
  const {
    onInit,
    swipeState: { dir, offset },
  } = useSwipe('low');

  // Effects to handle swipe gestures.
  useEffect(() => {
    if (
      (dir === 'RIGHT' && direction === 'horizontal') ||
      (dir === 'BOTTOM' && direction === 'vertical')
    ) {
      handlePrevious();
    } else if (
      (dir === 'LEFT' && direction === 'horizontal') ||
      (dir === 'TOP' && direction === 'vertical')
    ) {
      handleNext();
    }
  }, [offset, dir]);

  // Effect to clean up on unmount.
  useEffect(() => {
    return () => {
      resizeObserver.current?.disconnect();
    };
  }, []);

  // Render the Carousel component.
  return (
    <div className={carouselContainerClass} ref={onInit}>
      <div
        className={wrapperClass}
        ref={onInitRef}
        style={carouselWrapperStyle}
      >
        <CarouselItems
          carouselItems={carouselItems}
          direction={direction}
          activePage={activePage}
          totalItems={trackCount.current}
          height={slideHeight}
          width={slideWidth}
        >
          {children}
        </CarouselItems>
      </div>
      <div className={carouselTrackClass}>
        {!isAutoPlaying && (
          <CarouselTrack
            activeIndex={activePage}
            handleSelection={handleActivatePage}
            length={trackCount.current}
            direction={direction}
            onNext={handleNext}
            onPrevious={handlePrevious}
            hideNext={hideNextButton}
            hidePrevious={hidePreviousButton}
            focusable={focusable}
            size={size}
          />
        )}
      </div>
    </div>
  );
};

Carousel.displayName = 'Carousel';

export { Carousel };
