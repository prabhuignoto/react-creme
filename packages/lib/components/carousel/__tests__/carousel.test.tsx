import { render, waitFor } from '@testing-library/react';
import React from 'react';
import { Carousel } from '../carousel';

describe('Carousel', () => {
  it('should render carousel', async () => {
    const { getByText } = render(
      <Carousel direction="horizontal">
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    await waitFor(() => {
      expect(getByText('one')).toBeInTheDocument();
      expect(getByText('two')).toBeInTheDocument();
    });
  });

  it('should render carousel snapshot', async () => {
    const { container } = render(
      <Carousel direction="horizontal">
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    await waitFor(
      () => {
        expect(container.firstChild).toMatchSnapshot();
      },
      { timeout: 1000 }
    );
  });

  it('should render vertical carousel', async () => {
    const { getByText } = render(
      <Carousel direction="vertical">
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    await waitFor(
      () => {
        expect(getByText('one')).toBeInTheDocument();
        expect(getByText('two')).toBeInTheDocument();
      },
      { timeout: 2000 }
    );
  });
});
