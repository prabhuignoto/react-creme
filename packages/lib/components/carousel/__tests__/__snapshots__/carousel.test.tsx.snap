// Vitest Snapshot v1

exports[`Carousel > should render carousel snapshot 1`] = `
<div
  class="rc-carousel-container rc-carousel-container-horizontal"
>
  <div
    class="rc-carousel-wrapper rc-carousel-wrapper-horizontal"
    style="--min-height: 400px; --transition: cubic-bezier(0.55, 0.08, 0.68, 0.53);"
  >
    <ul
      class="rc-carousel"
      role="list"
      style="transform: translateX(-0px); width: 0px;"
    >
      <li
        aria-hidden="false"
        class="rc-carousel-item"
        data-visible="true"
        role="listitem"
        style="top: 0px; visibility: hidden; width: 0px;"
      >
        <div
          class="rc-carousel-item-container"
        >
          <span>
            one
          </span>
        </div>
      </li>
      <li
        aria-hidden="true"
        class="rc-carousel-item"
        data-visible="false"
        role="listitem"
        style="top: 0px; visibility: hidden; width: 0px;"
      >
        <div
          class="rc-carousel-item-container"
        >
          <span>
            two
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="rc-carousel-track-wrapper rc-carousel-track-wrapper-horizontal"
  >
    <div
      class="rc-carousel-track-wrapper rc-carousel-track-wrapper-horizontal"
    >
      <span
        class="rc-carousel-btn rc-carousel-btn-left rc-carousel-btn-horizontal rc-carousel-btn-hide"
      >
        <button
          aria-label="Previous"
          class="rc-btn-sm rc-btn-icon rc-btn"
          role="button"
          style="outline: none; position: relative;"
          tabindex="0"
        >
          <span
            class="rc-btn-icon-container"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 18l6-6-6-6"
              />
            </svg>
          </span>
          <span
            class="rc-focus-ring"
            style="width: 6px; height: 6px;"
          />
        </button>
      </span>
      <ul
        class="rc-carousel-track rc-carousel-track-horizontal"
        role="list"
      >
        <li
          class="rc-carousel-track-item rc-carousel-track-item-selected"
          role="listitem"
        />
        <li
          class="rc-carousel-track-item"
          role="listitem"
        />
      </ul>
      <span
        class="rc-carousel-btn rc-carousel-btn-right rc-carousel-btn-horizontal"
      >
        <button
          aria-label="Next"
          class="rc-btn-sm rc-btn-icon rc-btn"
          role="button"
          style="outline: none; position: relative;"
          tabindex="0"
        >
          <span
            class="rc-btn-icon-container"
          >
            <svg
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 18l6-6-6-6"
              />
            </svg>
          </span>
          <span
            class="rc-focus-ring"
            style="width: 6px; height: 6px;"
          />
        </button>
      </span>
    </div>
  </div>
</div>
`;
